File: java/src/test/java/com/example/QueueTest.java
package com.example;

import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

public class QueueTest extends TestCase {
    public QueueTest(String testName) {
        super(testName);
    }

    public static Test suite() {
        return new TestSuite(QueueTest.class);
    }

    private final Struct[] values = {
            new Struct(2),
            new Struct(4),
            new Struct(6),
            new Struct(8),
            new Struct(10),
    };

    public void testQueueAdd() {
        Queue<Struct> queue = new Queue<>();

        for (int i = 0; i < values.length; i++) {
            Struct val = values[i];
            queue.add(val);

            assertEquals(queue.size(), i + 1);
        }

        for (int i = 0; i < values.length; i++) {
            Struct val = values[i];
            Struct v = queue.remove().get();
            assertEquals(val.val, v.val);
        }
        assertEquals(queue.remove().isEmpty(), true);
    }

    public void testGetFromArray() {
        Queue<Struct> queue = new Queue<>(values);

        for (int i = 0; i < values.length; i++) {
            Struct val = values[i];
            Struct v = queue.remove().get();
            assertEquals(val.val, v.val);
        }
        assertEquals(queue.remove().isEmpty(), true);
    }

    public void testPeekElement() {
        Queue<Struct> queue = new Queue<>(values);
        for (int i = 0; i < values.length; i++) {
            Struct val = values[i];
            Struct v = queue.peek().get();
            assertEquals(val.val, v.val);
            queue.remove();
        }
        assertEquals(queue.remove().isEmpty(), true);
    }

    public void testListToString() {
        Queue<Struct> queue = new Queue<>(values);

        String str = queue.toString((Struct s) -> String.valueOf(s.val));
        assertEquals(str, "[2,4,6,8,10]");

        str = queue.toString((Struct s) -> String.valueOf(s.val) + "0");
        assertEquals(str, "[20,40,60,80,100]");
    }
}

class Struct {
    public int val;

    public Struct(int value) {
        this.val = value;
    }
}

File: python/queue_test.py
import unittest
from queue_list import Queue

values = [
    {'val': 2},
    {'val': 4},
    {'val': 6},
    {'val': 8},
    {'val': 10},
]

class TestQueue(unittest.TestCase):
    def test_queue_add(self):
        global values
        queue = Queue()

        for i, v in enumerate(values):
            queue.add(v)
            self.assertEqual(queue.size(), i+1)
        
        for i, val in enumerate(values):
            v = queue.remove()
            self.assertEqual(val['val'], v['val'])


    def test_get_from_array(self):
        global values
        queue = Queue(values)

        for i, val in enumerate(values):
            v = queue.remove()
            self.assertEqual(val['val'], v['val'])


    def test_peek_element(self):
        global values
        queue = Queue(values)

        for i, val in enumerate(values):
            v = queue.peek()
            self.assertEqual(val['val'], v['val'])
            queue.remove()
        self.assertEqual(queue.remove(), None)


    def test_to_string(self):
        global values
        queue = Queue(values)
        s = queue.to_string(lambda x: str(x['val']))
        self.assertEqual(s, '[2,4,6,8,10]')


if __name__ == '__main__':
    unittest.main()
